### PR TITLE
Fix shulkers teleporting on the road

### PR DESCRIPTION
**Pull request by NotMyFault**
## Overview

Fixes https://github.com/IntellectualSites/PlotSquared/issues/3491

## Submitter Checklist

- [X] Make sure you are opening from a topic branch (**/feature/fix/docs/ branch** (right side)) and not your main branch.
- [X] Ensure that the pull request title represents the desired changelog entry.
- [X] I tested my changes and approved their functionality.
- [X] I ensured my changes do not break other parts of the code
- [X] I read and followed the [contribution guidelines](https://github.com/IntellectualSites/.github/blob/main/CONTRIBUTING.md)

## Maintainer checklist

Before the changes are marked as `Ready for merge`: 

- [ ] There are at least 2 approvals from project developers or maintainers for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [X] Changelog entries in the PR title are correct
- [X] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional modification steps from users, e.g. for permission changes, make sure the entry is added to the pull request description so it can be appended to the changelog.

### CODE REVIEW

**Review by PierreSchwang — COMMENTED** [review]
Looks good so far, but wouldn't it be better to set a new location inside the plot instead of cancelling the whole event?

**Comment by NotMyFault** [comment]
> Looks good so far, but wouldn't it be better to set a new location inside the plot instead of cancelling the whole event?

The teleport to a block that is not within the plot is cancelled, which leads to the shulker searching for a new location inside the plot.

**Review by PierreSchwang — APPROVED** [review]

File: Bukkit/src/main/java/com/plotsquared/bukkit/listener/EntitySpawnListener.java [hunk 1]
```diff
@@ -191,8 +192,32 @@ public void onVehicle(VehicleBlockCollisionEvent event) {
...
+        Location toLocLocation = BukkitUtil.adapt(toLocation.getLocation());
+        PlotArea area = toLocLocation.getPlotArea();
+
+        if (area == null) {
+            if (fromLocation.getType() == EntityType.SHULKER && fromArea != null) {
+                event.setCancelled(true);
+            }
+            return;
+        }
+        Plot plot = area.getOwnedPlot(toLocLocation);
```
Comment (yannicklamprecht) [inline Bukkit/src/main/java/com/plotsquared/bukkit/listener/EntitySpawnListener.java:209]: I would rename it to plotTo, to be consistent.

File: Bukkit/src/main/java/com/plotsquared/bukkit/listener/EntitySpawnListener.java [hunk 2]
```diff
@@ -191,8 +192,32 @@ public void onVehicle(VehicleBlockCollisionEvent event) {
...
     public void onTeleport(EntityTeleportEvent event) {
-        Entity ent = event.getEntity();
-        if (ent instanceof Vehicle || ent instanceof ArmorStand) {
+        Entity entity = event.getEntity();
+        Entity fromLocation = event.getEntity();
+        Block toLocation = event.getTo().getBlock();
+        final Location fromLocLocation = BukkitUtil.adapt(fromLocation.getLocation());
+        final PlotArea fromArea = fromLocLocation.getPlotArea();
+        Location toLocLocation = BukkitUtil.adapt(toLocation.getLocation());
+        PlotArea area = toLocLocation.getPlotArea();
```
Comment (yannicklamprecht) [inline Bukkit/src/main/java/com/plotsquared/bukkit/listener/EntitySpawnListener.java:201]: toArea for consistency.

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]
Just some inconsistent variable naming I stumbled over.

**Review by yannicklamprecht — APPROVED** [review]

**Review by dordsor21 — APPROVED** [review]

**Comment by NotMyFault** [comment]
This PR is ready to be merged, we will merge it in ~24 hours if there is no negative feedback, as documented by the [merge process](https://github.com/IntellectualSites/IntellectualSites/blob/main/docs/MAINTAINERS_CONTRIBUTORS.adoc#merge-process) documentation.

**Pull request merged**